### PR TITLE
Remove anonymous functions from sheaves

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
Also fixes the issue pointed out by @simonbrandhorst on #4110 .

### CODE REVIEW

**Review by afkafkafk13 — COMMENTED** [review]
I had always been confused by your production_function and restriction_function arguments. This is now a lot clearer (thanks to the needs of serialization efforts).

Overall this looks good, but moving stuff at the same time as making this change was not overly helpful for reviewing.

**Comment by afkafkafk13** [comment]
Let me know, wether you are still adding stuff to this or whether you would like a complete review.

**Comment by HechtiDerLachs** [comment]
Thanks for having a look. Unfortunately, moving the functions was unavoidable. Yes, a full review would be appreciated. Nothing changed about the internal mathematics, though.

File: src/AlgebraicGeometry/Schemes/Sheaves/Sheaves.jl [hunk 1]
```diff
@@ -59,7 +59,7 @@ function produce_object(F::AbsPreSheaf, U)
 end
 
 ### temporary workaround
-produce_object(F::AbsPreSheaf, U::AbsAffineScheme) = production_func(F)(F, U)
+produce_object(F::AbsPreSheaf, U::AbsAffineScheme) = produce_object(underlying_presheaf(F), U)
```
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Sheaves/Sheaves.jl]: Is the comment about the temporary workaround still valid? If not, please change.

File: src/AlgebraicGeometry/Schemes/Sheaves/Types.jl [hunk 2]
```diff
@@ -44,19 +44,15 @@ A basic minimal implementation of the interface for `AbsPreSheaf`; to be used in
 
   # production functions for new objects
   is_open_func::Function # To check whether one set is open in the other
-  production_func::Union{Function, Nothing} # To produce ℱ(U) for U ⊂ X
-  restriction_func::Union{Function, Nothing} # To produce the restriction maps ℱ(U) → ℱ(V) for V ⊂ U ⊂ X open
```
Comment (afkafkafk13) [inline src/AlgebraicGeometry/Schemes/Sheaves/Types.jl]: here again: please check whether the comment still reflects the current situation and change, if necessary. We do not want to confuse programmers who take a look at this as a model.

**Review by afkafkafk13 — APPROVED** [review]
I marked two places where comments should be changed to reflect the current implementation. Otherwise this is good to go.

**Pull request merged**